### PR TITLE
Revert & fix #9690 workarounds

### PR DESCRIPTION
**Pull request by valeriosetti**
## Description

This PR fixes some workaround which were introduced in #9690:

- re-enable all the tests in `test_suite_pkwrite.data`
- use `-O2` flag instead of `-O3` only for old (i.e. version < 7.0) GCC compilers

Depends on #9690

## PR checklist

- [ ] **changelog** not required because: 
- [ ] **development PR** not required because it's this one
- [ ] **framework PR** not required because that repo is not affected by this PR
- [ ] **3.6 PR** https://github.com/Mbed-TLS/mbedtls/pull/9728
- [ ] **2.28 PR** not required
- [] **tests**  not required

### CODE REVIEW

File: tf-psa-crypto/tests/suites/test_suite_pkwrite.data
Comment (gilles-peskine-arm) [inline tf-psa-crypto/tests/suites/test_suite_pkwrite.data]: You'll need to add a signoff line on the revert commit, otherwise the DCO check complains.

(Please wait until #9690 is merged to rebase.)

**Review by gilles-peskine-arm — APPROVED** [review]
Thanks, LGTM

**Comment by valeriosetti** [comment]
@davidhorstmann-arm if you are OK with the changes introduced by this PR I will start the backport of it. Wdyt?

**Pull request merged**
